plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftLaunchController.java

```diff
@@ -50,20 +67,83 @@ protected static IControllableServerBehavior getServerBehavior(ILaunchConfigurat
 	@Override
 	public void launch(ILaunchConfiguration configuration, String mode,
 			ILaunch launch, IProgressMonitor monitor) throws CoreException {
-		IControllableServerBehavior beh = getServerBehavior(configuration);
-		if( beh != null ) {
-			((ControllableServerBehavior)beh).setServerStarting();
-			int state = pollState();
-			if( state == IServer.STATE_STARTED) {
-				((ControllableServerBehavior)beh).setServerStarted();
-				((ControllableServerBehavior)getControllableBehavior()).setRunMode(mode);
-			} else {
-				((ControllableServerBehavior)beh).setServerStopped();
-				((ControllableServerBehavior)getControllableBehavior()).setRunMode(null);
-			}
+		IControllableServerBehavior serverBehavior = getServerBehavior(configuration);
+		if( !(serverBehavior instanceof ControllableServerBehavior )) {
+			throw toCoreException("Unable to find a IControllableServerBehavior instance");
+		}
+		ControllableServerBehavior beh = (ControllableServerBehavior) serverBehavior;
+		
+		beh.setServerStarting();
+		
+		IServer server = beh.getServer();
+		
+		IDeploymentConfig dc = OpenShiftServerUtils.getDeploymentConfig(server);
+		DebuggingContext debugContext = OpenShiftDebugUtils.get().getDebuggingContext(dc);
+		if( "debug".equals(mode)) {
+			startDebugging(server, dc, debugContext, monitor);
+		} else {//run, profile
+			stopDebugging(dc, debugContext, monitor);
+		}
+		
+		int state = pollState();
+		if( state == IServer.STATE_STARTED) {
```

Comment:
No need to synchronize anything here.  It's our job to set the server state here. The IServer is waiting for us to beh.setServerStarted() on success or beh.setServerStopped() on a startup-fail.   Since our initial poll indicates it's already up, we set it to started.